### ISSUE TITLE
Issue with "Drawing graphics": (A simple character animation)

### ISSUE THREAD

**Issue by hustnzj** [mdn/content#8448]
MDN URL: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Drawing_graphics

#### What information was incorrect, unhelpful, or incomplete?

> This effectively means that we are only updating the sprite on every 13th frame, or roughly about 5 frames a second...

#### Specific section or headline?
A simple character animation

#### What did you expect to see?
I think it's better to change `posX % 13` to `posX % 20`.

#### Did you test this? If so, how?
Yes.
* First of all, this function will execute 60 times per second on my screen - see the `console.log('hh')` and the `setInterval` line. 
* If this number is 13, the condition `posX % 13 === 0` will only be true when posX is one of the multipliers of 26 - 0,26,52,78,104,130 as the posX variable increments by 2 each frame. 
* That's to say every 13 frames the spirit would change. But the posX only increment 120 in one second. So if this number is 20, the condition will be true at 0,20,40,60,80,120 and the spirit will be exactly changed 5 times in one second.

The test code as below:
```html
<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>A simple character animation</title>
  <style>
    body {
      margin: 0;
      overflow: hidden;
    }
  </style>
</head>

<body>
  <canvas class="myCanvas">
    <p>Add suitable fallback here.</p>
  </canvas>

  <script>
    const canvas = document.querySelector('.myCanvas');
    const width = canvas.width = window.innerWidth;
    const height = canvas.height = window.innerHeight;
    const ctx = canvas.getContext('2d');

    ctx.fillStyle = 'rgb(0,0,0)';
    ctx.fillRect(0, 0, width, height);
    ctx.translate(width / 2, height / 2)

    let image = new Image()
    image.src = 'walk-right.png'
    image.onload = draw

    setInterval(() => {
      console.log('1s passed')
    }, 1000)

    let sprite = 0
    let posX = 0

    function draw() {
      ctx.fillRect(-width / 2, -height / 2, width, height)
      ctx.drawImage(image, sprite * 102, 0, 102, 148, 0 + posX, -74, 102, 148)

      console.log('hh')
      if (posX > width / 2) {
        newStartPos = -(width / 2 + 102)
        posX = Math.ceil(newStartPos)
        console.log(posX)
      } else {
        posX += 2
      }
      // 0 120
      // 0 26 52 78 104 130
      // 0 20 40 60 80 120
      if (posX % 20 === 0) {
        if (sprite === 5) {
          sprite = 0
        } else {
          sprite++
        }
      }
      window.requestAnimationFrame(draw)
    }
  </script>
</body>

</html>
```

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/learn/javascript/client-side_web_apis/drawing_graphics`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Drawing_graphics
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/javascript/client-side_web_apis/drawing_graphics/index.html
* Last commit: https://github.com/mdn/content/commit/3af43c1dcfb9be19c73cf3fdd1178db9c74aa064
* Document last modified: 2021-04-27T01:03:07.000Z

</details>

**Comment by vlwkaos** [comment]
@sideshowbarker This requires change in both [learning-area](https://github.com/mdn/learning-area) code sample and page content. I am fairly new to mdn repositories, would you mind explaining to me how cases like these are handled?

**Comment by sideshowbarker** [comment]
> @sideshowbarker This requires change in both [learning-area](https://github.com/mdn/learning-area) code sample and page content. I am fairly new to mdn repositories, would you mind explaining to me how cases like these are handled?

Looking at https://github.com/mdn/learning-area/pull/396 and https://github.com/mdn/content/pull/8535, it seems like you made the changes in the appropriate places. The way those will get handled is that they’ll be reviewed and merged separately, and deployed separately.

But part of what’s involved in resolving MDN issues is confirming that what’s asserted in a particular issue description is actually correct.

So can you explain what steps you took to confirm that what’s asserted in this issue description is correct?

For example, did you run a test and observe that with `posX % 13` the sprite doesn’t get updated every 5 seconds, but that with `posX % 20` it does?

What’s asserted in the issue description *sounds* correct to me — but without actually testing it myself, I personally did not feel confident enough about it to write up pull request myself with the change.

However, I don’t really really have a lot of familiarity with using the 2context API. But I could it’s possible you have enough familiarity with using the 2context API to be confident that what the issue description is asserting is true.

**Comment by vlwkaos** [comment]
> But part of what’s involved in resolving MDN issues is confirming that what’s asserted in a particular issue description is actually correct.

@sideshowbarker 
Indeed! Thanks for a kind reply.

After a further testing, I noticed that changing the code to `posX % 20` do update the sprite 5 times per second.

However, there was a caveat due to how `posX` value changes after the character reaches the end of the screen. 

```javascript
newStartPos = -((width/2) + 102);
```

Apparently, using an even number to update sprite per frame won't work with even number of canvas width since the value of `newStartPos` becomes odd. I could edit those part of the code to make it work. 

But on my second thought, the changes asserted in the issue description is not a necessary change due to:

> This effectively means that we are only updating the sprite on every 13th frame, or roughly about 5 frames a second...

1. The original content description acknowledges that it is roughly about 5 frames a second, which is a correct information.
2. In this case, sprite updating exactly 5 times a second is not a necessary change, since the purpose of `posX%{frames}` is to relatively update the sprite slower than every 60 frame, not to update the sprite exactly 5 times a second.

So sprite not updating 5 times a second does not look like an issue to me.

If you think my reasoning is ok, it would be nice of you to close the PRs that I made. (Or should I do it..?)

**Comment by sideshowbarker** [comment]
> If you think my reasoning is ok, it would be nice of you to close the PRs that I made. (Or should I do it..?)

I agree with that reasoning, so I’m going ahead and closing #8535, and also closing this issue.

As far as https://github.com/mdn/learning-area/pull/396, I don’t actually have the permissions needed to close PRs in that repo, so please go ahead and close that one on your own.

Thanks again very much for investigating this issue.
